contracts/facets/YGFacetZaynFi.sol

```diff
@@ -119,4 +122,110 @@ contract YGFacetZaynFi is IYGFacetZaynFi {
+
+        // Start looping through each combination
+        for (uint i; i < length; ) {
+            uint termId = termIds[i];
+            address user = users[i];
+
+            LibYieldGenerationStorage.YieldGeneration storage yield = LibYieldGenerationStorage
+                ._yieldStorage()
+                .yields[termId];
+
+            // Make sure user is part of this term and has enabled yield generation
+            require(yield.hasOptedIn[user], "User not part of yield generation");
+
+            // Zaynfi's addresses
+            address vaultAddress = yield.providerAddresses["ZaynVault"];
+            address zapAddress = yield.providerAddresses["ZaynZap"];
+
+            // Calculate what each user is owed
+            int reimbursement = _calculateReimbursement(originalWithdrawals[i], originalShares[i], yield);
+
+            if (reimbursement > 0) {
+                // Reimbursement is positive, this means the user withdrew less shares than he was supposed to
+                uint neededShares = uint(reimbursement);
+
+                // Code copied from _withdrawYG, get the amount of shares back and give it to the user
+                uint withdrawnYield = IZaynZapV2TakaDAO(zapAddress).zapOutETH(
+                    vaultAddress,
+                    neededShares,
+                    termId
+                );
+
+                yield.withdrawnYield[user] += withdrawnYield;
+                yield.availableYield[user] += withdrawnYield;
+
+                // Claim the yield right away and send it to the user
+                LibYieldGeneration._claimAvailableYield(termId, user, user);
+                
+            } else if (reimbursement < 0) {
+                // When there is a negative reimbursement, we compensate the pool by adding back the exact amount of shares that were lost
+                uint neededShares = uint(reimbursement * -1);
+
+                // Calculate the amount of eth we need to deposit to get the desired shares
+                uint pricePerShare = IZaynVaultV2TakaDao(vaultAddress)
+                    .getPricePerFullShare();
+
+                uint neededEth = neededShares * pricePerShare;
+                uint sharesBefore = IZaynVaultV2TakaDao(vaultAddress).balanceOf(termId);
+
+                // Make sure we have enough eth
+                require(neededEth + usedValue <= msg.value, "Not enough ETH value sent");
+
+                // Deposit the amount of shares we lost
+                IZaynZapV2TakaDAO(zapAddress).zapInEth{value: neededEth}(
+                    vaultAddress,
+                    termId
+                );
+
+                // Increment the used value so far
+                usedValue += neededEth;
+
+                // Validate the amount of shares deposited
+                uint sharesAfter = IZaynVaultV2TakaDao(vaultAddress).balanceOf(termId);
+                require(neededShares == (sharesAfter - sharesBefore), "Invalid amount of shares deposited");
```

Comment:
There is a little difference that fails here. For example for the first user in the inputs the neededShares are 158058370216700 and the sharesAfter - sharesBefore are 157458744451636

Reply:
What we can do is we should ZapIn 2 times as much eth, then zapput the difference in before and after so we have an exact amount of shares zapped in